<a href="https://colab.research.google.com/github/reesha-rsh/MLb4/blob/main/Homework%5CHomework_2_0_Pandas_Toxic_Comments_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
#basics
import pandas as pd
import numpy as np

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [100]:
train = pd.read_csv("/content/drive/MyDrive/MLb4/Homework/2.0 Pandas - Toxic Comments EDA/train_1.csv")

In [4]:
train.head()

Page  2015-07-01  2015-07-02  \
0            2NE1_zh.wikipedia.org_all-access_spider        18.0        11.0   
1             2PM_zh.wikipedia.org_all-access_spider        11.0        14.0   
2              3C_zh.wikipedia.org_all-access_spider         1.0         0.0   
3         4minute_zh.wikipedia.org_all-access_spider        35.0        13.0   
4  52_Hz_I_Love_You_zh.wikipedia.org_all-access_s...         NaN         NaN   

   2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  2015-07-08  \
0         5.0        13.0        14.0         9.0         9.0        22.0   
1        15.0        18.0        11.0        13.0        22.0        11.0   
2         1.0         1.0         0.0         4.0         0.0         3.0   
3        10.0        94.0         4.0        26.0        14.0         9.0   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   2015-07-09  ...  2016-12-22  2016-12-23  2016-12-24  2016-12-25  \
0        26.0  ...        32.0        63.0        15.0        26.0   
1        10.0  ...        17.0        42.0        28.0        15.0   
2         4.0  ...         3.0         1.0         1.0         7.0   
3        11.0  ...        32.0        10.0        26.0        27.0   
4         NaN  ...        48.0         9.0        25.0        13.0   

   2016-12-26  2016-12-27  2016-12-28  2016-12-29  2016-12-30  2016-12-31  
0        14.0        20.0        22.0        19.0        18.0        20.0  
1         9.0        30.0        52.0        45.0        26.0        20.0  
2         4.0         4.0         6.0         3.0         4.0        17.0  
3        16.0        11.0        17.0        19.0        10.0        11.0  
4         3.0        11.0        27.0        13.0        36.0        10.0  

[5 rows x 551 columns]

In [76]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145063 entries, 0 to 145062
Columns: 551 entries, Page to 2016-12-31
dtypes: float64(550), object(1)
memory usage: 609.8+ MB


In [38]:
import re
import string

In [15]:
#train["PageName"] = train["Page"].str.split(".", expand = True)[0]

In [31]:
#train["Language"] = train["PageName"].apply(lambda x: (re.split("_",str(x)))[-1])

In [80]:
#train["Language"]  = train["Page"].apply(lambda x: (re.findall("_\w{2}\.",str(x))))

In [109]:
train[(train["Page"].str.contains("zh")==True) & (train["Page"].str.contains("wikipedia")==False)]

Page  2015-07-01  \
13605  Category:Mridanga_Saileswari_Kshethram,_Muzhak...         NaN   
13706  Category:Reply_of_the_Zaporozhian_Cossacks_(Ru...         1.0   
13996  File:Brezhnev_LI_Pasport_1947.jpg_commons.wiki...         2.0   
14318  File:Pleven_TodorBozhinov_(5).jpg_commons.wiki...         NaN   
19628  API:Main_page/zh_www.mediawiki.org_all-access_...        32.0   
...                                                  ...         ...   
84274  Manual:Configuration_settings/zh_www.mediawiki...         4.0   
84302  Manual:Installation_guide/zh_www.mediawiki.org...         4.0   
84602  Special:WhatLinksHere/Help:Categories/zh_www.m...         NaN   
84627  Topic:Szhapu3bcbotvq3l_www.mediawiki.org_all-a...         NaN   
84632  Topic:T37tizherg285uml_www.mediawiki.org_all-a...         NaN   

       2015-07-02  2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  \
13605         NaN         NaN         NaN         0.0         0.0         NaN   
13706         0.0         0.0         1.0         1.0         0.0         1.0   
13996         1.0         3.0         1.0         3.0         3.0         3.0   
14318         NaN         2.0         1.0         NaN         NaN         NaN   
19628        40.0        18.0        19.0        35.0        30.0        28.0   
...           ...         ...         ...         ...         ...         ...   
84274         2.0         2.0         5.0         1.0         2.0         1.0   
84302        13.0        14.0        19.0        25.0        28.0        20.0   
84602         NaN         NaN         NaN         1.0         NaN         NaN   
84627         NaN         NaN         NaN         NaN         NaN         NaN   
84632         NaN         NaN         NaN         NaN         NaN         NaN   

       2015-07-08  2015-07-09  ...  2016-12-22  2016-12-23  2016-12-24  \
13605         NaN         NaN  ...         0.0         2.0         0.0   
13706         1.0         1.0  ...         2.0         3.0         0.0   
13996         5.0         7.0  ...         0.0         1.0         3.0   
14318         0.0         NaN  ...         3.0         1.0         NaN   
19628        19.0         9.0  ...        41.0        54.0        60.0   
...           ...         ...  ...         ...         ...         ...   
84274         0.0         1.0  ...         5.0         8.0         9.0   
84302         5.0         1.0  ...         3.0        10.0        11.0   
84602         NaN         NaN  ...         NaN         0.0         0.0   
84627         NaN         NaN  ...         NaN         1.0         0.0   
84632         NaN         NaN  ...         NaN         0.0         1.0   

       2016-12-25  2016-12-26  2016-12-27  2016-12-28  2016-12-29  2016-12-30  \
13605         NaN         0.0         0.0         1.0         2.0         0.0   
13706         2.0         1.0         1.0         0.0         1.0         1.0   
13996         1.0         1.0         2.0         2.0         3.0         2.0   
14318         1.0         NaN         2.0         NaN         2.0         3.0   
19628        37.0        52.0        36.0        41.0        22.0        42.0   
...           ...         ...         ...         ...         ...         ...   
84274         9.0         7.0         3.0         3.0         3.0         4.0   
84302         9.0         9.0        10.0         8.0         3.0         6.0   
84602         0.0         1.0         NaN         NaN         NaN         0.0   
84627         1.0         NaN         NaN         2.0         NaN         NaN   
84632         0.0         NaN         NaN         NaN         NaN         NaN   

       2016-12-31  
13605         0.0  
13706         1.0  
13996         1.0  
14318         NaN  
19628        35.0  
...           ...  
84274         5.0  
84302         6.0  
84602         NaN  
84627         2.0  
84632         NaN  

[134 rows x 551 columns]

In [114]:
train["Language"] = train["Page"].apply(lambda x: (re.search("_(\w{2})\.wikipedia\.org",str(x)))).apply(lambda x: str(x.group(1)) if x else None )

In [111]:
train[train["Language"].isnull()]


Page  2015-07-01  \
13332    Accueil_commons.wikimedia.org_all-access_spider        55.0   
13333  Atlas_of_Asia_commons.wikimedia.org_all-access...         5.0   
13334  Atlas_of_Europe_commons.wikimedia.org_all-acce...         4.0   
13335  Atlas_of_World_War_II_commons.wikimedia.org_al...         5.0   
13336  Atlas_of_colonialism_commons.wikimedia.org_all...         8.0   
...                                                  ...         ...   
84730  help:contents_www.mediawiki.org_all-access_spider         NaN   
84731  how_to_contribute_www.mediawiki.org_all-access...         0.0   
84732      mediawiki_www.mediawiki.org_all-access_spider         0.0   
84733   special:book_www.mediawiki.org_all-access_spider         0.0   
84734  special:translate_www.mediawiki.org_all-access...         0.0   

       2015-07-02  2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  \
13332        48.0        44.0        35.0        46.0        41.0        32.0   
13333         3.0         4.0         6.0         3.0         1.0         5.0   
13334         6.0         9.0         7.0         6.0         4.0         5.0   
13335         2.0         6.0         6.0         8.0        11.0         5.0   
13336         6.0        15.0         5.0         5.0         9.0        10.0   
...           ...         ...         ...         ...         ...         ...   
84730         0.0         NaN         NaN         0.0         0.0         NaN   
84731         0.0         0.0         0.0         0.0         0.0         0.0   
84732         0.0         0.0         0.0         0.0         0.0         0.0   
84733         0.0         0.0         0.0         0.0         NaN         0.0   
84734         0.0         0.0         0.0         0.0         1.0         NaN   

       2015-07-08  2015-07-09  ...  2016-12-23  2016-12-24  2016-12-25  \
13332        43.0        40.0  ...        74.0        34.0        62.0   
13333         6.0         6.0  ...        43.0        44.0        38.0   
13334         6.0        13.0  ...        44.0        47.0        46.0   
13335         5.0         6.0  ...        18.0        13.0        11.0   
13336         3.0         9.0  ...        32.0        40.0        38.0   
...           ...         ...  ...         ...         ...         ...   
84730         NaN         0.0  ...         7.0         8.0        15.0   
84731         0.0         0.0  ...        37.0         4.0        18.0   
84732         0.0         0.0  ...        48.0        13.0        24.0   
84733         0.0         0.0  ...        40.0         3.0        53.0   
84734         NaN         0.0  ...        26.0         1.0        39.0   

       2016-12-26  2016-12-27  2016-12-28  2016-12-29  2016-12-30  2016-12-31  \
13332        88.0        57.0        74.0        61.0        51.0        60.0   
13333        31.0        29.0        37.0        32.0        29.0        34.0   
13334        35.0        31.0        31.0        34.0        28.0        27.0   
13335         8.0         8.0        13.0         5.0        11.0         6.0   
13336        29.0        19.0        30.0        29.0        26.0        26.0   
...           ...         ...         ...         ...         ...         ...   
84730         7.0        18.0        17.0        15.0         3.0        11.0   
84731         2.0       172.0        19.0        29.0         2.0       178.0   
84732        11.0        12.0        51.0        23.0         3.0        48.0   
84733       300.0         3.0        44.0         6.0         0.0       154.0   
84734       346.0         2.0        28.0         4.0       163.0         1.0   

       Language  
13332      None  
13333      None  
13334      None  
13335      None  
13336      None  
...         ...  
84730      None  
84731      None  
84732      None  
84733      None  
84734      None  

[17855 rows x 552 columns]

In [112]:
train.head()

Page  2015-07-01  2015-07-02  \
0            2NE1_zh.wikipedia.org_all-access_spider        18.0        11.0   
1             2PM_zh.wikipedia.org_all-access_spider        11.0        14.0   
2              3C_zh.wikipedia.org_all-access_spider         1.0         0.0   
3         4minute_zh.wikipedia.org_all-access_spider        35.0        13.0   
4  52_Hz_I_Love_You_zh.wikipedia.org_all-access_s...         NaN         NaN   

   2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  2015-07-08  \
0         5.0        13.0        14.0         9.0         9.0        22.0   
1        15.0        18.0        11.0        13.0        22.0        11.0   
2         1.0         1.0         0.0         4.0         0.0         3.0   
3        10.0        94.0         4.0        26.0        14.0         9.0   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   2015-07-09  ...  2016-12-23  2016-12-24  2016-12-25  2016-12-26  \
0        26.0  ...        63.0        15.0        26.0        14.0   
1        10.0  ...        42.0        28.0        15.0         9.0   
2         4.0  ...         1.0         1.0         7.0         4.0   
3        11.0  ...        10.0        26.0        27.0        16.0   
4         NaN  ...         9.0        25.0        13.0         3.0   

   2016-12-27  2016-12-28  2016-12-29  2016-12-30  2016-12-31  Language  
0        20.0        22.0        19.0        18.0        20.0        zh  
1        30.0        52.0        45.0        26.0        20.0        zh  
2         4.0         6.0         3.0         4.0        17.0        zh  
3        11.0        17.0        19.0        10.0        11.0        zh  
4        11.0        27.0        13.0        36.0        10.0        zh  

[5 rows x 552 columns]

1. Вікіпедії яких країн доступні в цьому датасеті?

In [120]:
list(filter(lambda item: item is not None, train["Language"].unique()))

['zh', 'fr', 'en', 'ru', 'de', 'ja', 'es']

Відповідь до завдання 1: **['zh', 'fr', 'en', 'ru', 'de', 'ja', 'es']**

*Думаю в завданні малось на увазі, на яких **мовах** доступні вікіпедії


2. Скільки сторінок з французької Вікіпедії у датасеті?

In [125]:
len(train[(train["Language"] == "fr")])

17802

Відповідь до завдання 2: **17802**

3. Вивести топ-3 статі для кожної країни за квітень 2016 року.
4. Знайти сторінку англійської вікіпедії з найбільшим середнім/медіаною/дисперсією за липень 2016 року.
5. Які статі отримали найбільший буст по переглядам за травень 2016 року в абсолютних показниках і відносних. (розглядати сторінки щонайменше з 1000 переглядами)
6. Візуалізувати кількість переглядів для топ-5 сторінок англійської Вікіпедії на одному графіку за перше півріччя 2016 року.
7. Порівняти розподіли німецьких і французьких сторінок вікіпедії за червень 2016 року.
8. Порівняти кількість переглядів для сторінок японської вікіпедії з розбивкою по дням тижня.